File: Utilities/Matlab/scripts/impinging_jet.m
% McDermott
% 12-8-2023
% impinging_jet.m

close all
clear all

figure
plot_style
set(gca,'Units',Plot_Units)
set(gca,'Position',[Plot_X Plot_Y Plot_Width Plot_Height])

outdir = '../../../out/Convection/';
pltdir = '../../Manuals/FDS_Verification_Guide/SCRIPT_FIGURES/';
chid = 'impinging_jet';

% plot the correlation

D_h = 0.2;                            % hydraulic diameter of the jet [m]
H   = 1;                              % distance from jet to wall [m]
mu  = 1.822E-05;                      % dynamic viscosity, [kg/m/s]
k   = 2.566E-02;                      % thermal conductivity [W/m/K]
Pr  = 0.71;                           % Prandtl number
T_j = 100;                            % tempetature of jet exit [C]
T_w = 20;                             % constant plate temperature [C]
rho_a = 1.200;                        % ambient density [kg/m3]
rho_j = rho_a * (T_w+273)/(T_j+273);  % jet fluid density [kg/m3]
nu = mu/rho_j;                        % kinematic viscosity [m2/s]
U_j = [10,40];                        % jet exit velocity [m/s]
A_r  = 0.01;                          % D_h^2/(4 r^2), where r defines the outer extent of the averaging region

% Plot correlation versus Re_j for a given geometry

% Martin corrleation
G = 2*sqrt(A_r)*( (1-2.2*sqrt(A_r))/(1+0.2*(H/D_h-6)*sqrt(A_r)) );

RE = linspace(2e3,5e5);
NU = G*2*sqrt(RE).*sqrt(1+0.005*RE.^0.55).* Pr^0.42;

K(1)=plot(RE,NU,'k-','linewidth',2); hold on

res_str = {'coarse','medium','fine'};
Re_str  = {'1e5','4e5'};

% relative error tolerance
E_tol = 0.1;

E_FDS = zeros(3,2);

for j=1:length(res_str)
    for i=1:length(Re_str)

        Re_j = D_h*U_j(i)/nu;  % jet Reynolds number

        Nu = G*2*sqrt(Re_j)*sqrt(1+0.005*Re_j^0.55) * Pr^0.42;

        h_cor = Nu * k/D_h;

        % FDS results

        M = importdata([outdir,'impinging_jet_Re_',Re_str{i},'_',res_str{j},'_devc.csv'],',',2);
        HF = mean(M.data(floor(end/2):end,find(strcmp(M.colheaders,'"HF"'))));

        qconv = HF * 1000; % W/m2
        h_fds = qconv/(T_j-T_w);
        Nu_fds = h_fds*D_h/k;

        E_FDS(i,j) = abs(Nu - Nu_fds)/abs(Nu);

        if E_FDS(i,j) > E_tol
          disp(['Matlab Warning: impinging jet error = ',num2str(E_FDS(i,j)),' at Re_j=',Re_str{i},', Res=',res_str{j}])
        end

        if i==1
            if j==1
                K(2)=plot(Re_j,Nu_fds,'bsq','linewidth',2); Key={'FDS coarse'};
            elseif j==2
                K(3)=plot(Re_j,Nu_fds,'rsq','linewidth',2); Key={'FDS coarse','FDS medium'};
            elseif j==3
                K(4)=plot(Re_j,Nu_fds,'gsq','linewidth',2); Key={'FDS coarse','FDS medium','FDS fine'};
            end
        elseif i==2
            if j==1
                plot(Re_j,Nu_fds,'bsq','linewidth',2)
            elseif j==2
                plot(Re_j,Nu_fds,'rsq','linewidth',2)
            elseif j==3
                plot(Re_j,Nu_fds,'gsq','linewidth',2)
            end
        end

    end
end

set(gca,'FontName',Font_Name)
set(gca,'FontSize',Label_Font_Size)

xlabel('Re','Interpreter',Font_Interpreter,'FontSize',Label_Font_Size)
ylabel('Nu','Interpreter',Font_Interpreter,'FontSize',Label_Font_Size)
lh=legend(K,['Martin',Key],'location','northwest');
set(lh,'FontName',Font_Name,'FontSize',Key_Font_Size);

Git_Filename = [outdir,'impinging_jet_Re_1e5_coarse_git.txt'];
addverstr(gca,Git_Filename,'linear')

set(gcf,'Visible',Figure_Visibility);
set(gcf,'Units',Paper_Units);
set(gcf,'PaperUnits',Paper_Units);
set(gcf,'PaperSize',[Paper_Width Paper_Height]);
set(gcf,'Position',[0 0 Paper_Width Paper_Height]);
print(gcf,'-dpdf',[pltdir,'impinging_jet_correlation']);


% plot the profile of heat transfer coefficient

style = {'k-.','k--','k-'};
clear K lh;

for i=1:length(Re_str)
    figure
    set(gca,'Units',Plot_Units)
    set(gca,'Position',[Plot_X Plot_Y Plot_Width Plot_Height])

    for j=1:length(res_str)

        M = importdata([outdir,'impinging_jet_Re_',Re_str{i},'_',res_str{j},'_line.csv'],',',2);
        x = M.data(:,1);
        q_x = M.data(:,find(strcmp(M.colheaders,'QCONV'))) * 1000;
        Nu_x = q_x/(T_j-T_w)*D_h/k;

        K(j)=plot(x,Nu_x,style{j}); hold on
    end
    xlabel('{\itx} (m)','Interpreter',Font_Interpreter,'FontSize',Label_Font_Size)
    ylabel('Nu_{Dh}({\itx})','Interpreter',Font_Interpreter,'FontSize',Label_Font_Size)

    axis([-.5 .5 0 1000])

    set(gca,'FontName',Font_Name)
    set(gca,'FontSize',Label_Font_Size)

    lh=legend(K,'D_h/\delta{\itx}=7','D_h/\delta{\itx}=14','D_h/\delta{\itx}=28','location','northwest');
    set(lh,'FontName',Font_Name,'FontSize',Key_Font_Size);

    Git_Filename = [outdir,'impinging_jet_Re_1e5_coarse_git.txt'];
    addverstr(gca,Git_Filename,'linear')

    set(gcf,'Visible',Figure_Visibility);
    set(gcf,'Units',Paper_Units);
    set(gcf,'PaperUnits',Paper_Units);
    set(gcf,'PaperSize',[Paper_Width Paper_Height]);
    set(gcf,'Position',[0 0 Paper_Width Paper_Height]);
    print(gcf,'-dpdf',[pltdir,'impinging_jet_local_Re_',Re_str{i}]);

    hold off
end

File: Utilities/Python/scripts/impinging_jet.py
#!/usr/bin/env python3
"""
impinging_jet.py
Converted from MATLAB script impinging_jet.m to Python.

Generates:
  - impinging_jet_correlation.pdf
  - impinging_jet_local_Re_1e5.pdf
  - impinging_jet_local_Re_4e5.pdf
"""

import os
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import fdsplotlib

plot_style = fdsplotlib.get_plot_style("fds")

# -----------------------------
# Paths and identifiers
# -----------------------------
firemodels_dir = os.path.normpath(os.path.join(os.path.dirname(__file__),'..','..','..','..'))
fds_dir = os.path.normpath(os.path.join(os.path.dirname(__file__),'..','..','..'))
outdir = os.path.join(firemodels_dir, 'out','Convection','')
pltdir = os.path.join(fds_dir, 'Manuals','FDS_Verification_Guide','SCRIPT_FIGURES','')
os.makedirs(pltdir, exist_ok=True)

chid = 'impinging_jet'

# -----------------------------
# Constants / inputs
# -----------------------------
D_h   = 0.2          # hydraulic diameter of the jet [m]
H     = 1.0          # distance from jet to wall [m]
mu    = 1.822e-05    # dynamic viscosity [kg/m/s]
k     = 2.566e-02    # thermal conductivity [W/m/K]
Pr    = 0.71         # Prandtl number [-]
T_j   = 100.0        # temperature of jet exit [C]
T_w   = 20.0         # constant plate temperature [C]
rho_a = 1.200        # ambient density [kg/m^3]
rho_j = rho_a * (T_w + 273.0) / (T_j + 273.0)  # jet fluid density [kg/m^3]
nu    = mu / rho_j   # kinematic viscosity [m^2/s]
U_j   = [10.0, 40.0] # jet exit velocities [m/s] corresponding to Re ~ 1e5 and 4e5
A_r   = 0.01         # D_h^2/(4 r^2), where r defines outer extent of the averaging region

res_list = ['coarse', 'medium', 'fine']
Re_str   = ['1e5', '4e5']

# Relative error tolerance for Nu comparison
E_tol = 0.1

# Get git hash
git_file = os.path.join(outdir, 'impinging_jet_Re_1e5_coarse_git.txt')
version_string = fdsplotlib.get_version_string(git_file)

# -----------------------------
# Plot 1: Martin correlation vs Re and FDS points
# -----------------------------
# Martin correlation coefficient G for given geometry
G = 2.0 * np.sqrt(A_r) * ((1 - 2.2*np.sqrt(A_r)) / (1 + 0.2*(H/D_h - 6)*np.sqrt(A_r)))

RE = np.linspace(2e3, 5e5, 1000)
NU = G * 2*np.sqrt(RE) * np.sqrt(1 + 0.005*RE**0.55) * (Pr**0.42)

fig = fdsplotlib.plot_to_fig(x_data=RE, y_data=NU, marker_style='k-', data_label='Martin',
                             x_min=0.0, x_max=5e5, y_min=0, y_max=600,
                             revision_label=version_string,
                             x_label='Re',
                             y_label='Nu')


markers = {'coarse': ('b', 's'), 'medium': ('r', 's'), 'fine': ('g', 's')}

E_FDS = np.zeros((2, 3))  # shape: (len(Re_str), len(res_list))

# Loop over resolutions and Re cases
for j, res in enumerate(res_list):
    for i, Re_tag in enumerate(Re_str):
        # Compute Re from U_j for this case
        Re_j = D_h * U_j[i] / nu

        # Martin Nu for this Re
        Nu_cor = G * 2*np.sqrt(Re_j) * np.sqrt(1 + 0.005*Re_j**0.55) * (Pr**0.42)

        # FDS results: read *_devc.csv and get average HF over second half
        devc_path = os.path.join(outdir, f'impinging_jet_Re_{Re_tag}_{res}_devc.csv')
        df = pd.read_csv(devc_path, header=1)
        
        # Mean HF over second half of samples
        n = len(df)
        hf_col = 'HF' if 'HF' in df.columns else df.columns[-1]  # simple fallback
        HF_mean = df[hf_col].iloc[n // 2 :].mean()

        qconv = HF_mean * 1000.0  # W/m^2
        h_fds = qconv / (T_j - T_w)
        Nu_fds = h_fds * D_h / k

        # relative error
        E = abs(Nu_cor - Nu_fds) / abs(Nu_cor)
        E_FDS[i, j] = E
        if E > E_tol:
            print(f'MATLAB Warning (ported): impinging jet error = {E:.3f} at Re_j={Re_tag}, Res={res}')

        label = f'FDS {res}' if i == 0 else None
        fig = fdsplotlib.plot_to_fig(x_data=Re_j, y_data=Nu_fds, data_label=label, figure_handle=fig, marker_style=markers[res], marker_fill_color=(1,1,1,0.0))

# Convert exponential to scientific
fdsplotlib.apply_global_exponent(fig.axes[0], axis='x', fontsize=plot_style['Label_Font_Size'])

# Save figure
corr_pdf = os.path.join(pltdir, 'impinging_jet_correlation.pdf')
fig.savefig(corr_pdf)
plt.close(fig)

# -----------------------------
# Plot 2: Local Nu(x) profiles for each Re case
# -----------------------------
styles = ['k-.', 'k--', 'k-']  # coarse, medium, fine

for i, Re_tag in enumerate(Re_str):
    fig = None
    for j, res in enumerate(res_list):
        line_path = os.path.join(outdir, f'impinging_jet_Re_{Re_tag}_{res}_line.csv')
        LINE = pd.read_csv(line_path, header=1)

        # x is first column; QCONV column used (match MATLAB)
        x = LINE.iloc[:, 0].to_numpy()
        q_col = 'QCONV' if 'QCONV' in LINE.columns else LINE.columns[-1]
        q_x = LINE[q_col].to_numpy() * 1000.0  # W/m^2

        Nu_x = (q_x / (T_j - T_w)) * D_h / k

        revlabel = version_string if j == 0 else None
        fig = fdsplotlib.plot_to_fig(x_data=x, y_data=Nu_x, marker_style=styles[j], data_label=r'$D_{h}/\delta x$ = %d'%(7*(2**j)),
                                     x_min=-0.5, x_max=0.5, y_min=0, y_max=1000, xnumticks=3, ynumticks=6,
                                     revision_label=revlabel, figure_handle=fig, 
                                     x_label='$x$ (m)',
                                     y_label=r'Nu$_{D_h}$($x$)', legend_location=2)

    fig.axes[0].xaxis.set_tick_params(pad=10)
    local_pdf = os.path.join(pltdir, f'impinging_jet_local_Re_{Re_tag}.pdf')
    fig.savefig(local_pdf)
    plt.close(fig)

print('Completed plots:')
print(' -', corr_pdf)
for Re_tag in Re_str:
    print(' -', os.path.join(pltdir, f'impinging_jet_local_Re_{Re_tag}.pdf'))
